src/components/Explore/filters/FilterRenderer.test.tsx

```diff
@@ -0,0 +1,127 @@
+import React from 'react';
+import { act, render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import { FilterRenderer, formatKeys, sortValues } from './FilterRenderer';
+import { AdHocVariableFilter, toOption } from '@grafana/data';
+import { FilterByVariable } from './FilterByVariable';
+
+jest.mock('utils/utils', () => ({
+  getTraceExplorationScene: () => ({
+    useState: jest.fn().mockReturnValue({ primarySignal: 'full_traces' }),
+    getMetricVariable: () => ({
+      useState: jest.fn().mockReturnValue('rate'),
+    }),
+  }),
+}));
+
+const mockFilter: AdHocVariableFilter = { key: 'testKey', operator: '=', value: 'testValue' };
+const mockKeys = [{ value: 'resource.tag1' }, { value: 'span.tag1' }, { value: 'kind' }, { value: 'resource.tag2' }];
+const mockValues = [toOption('value1'), toOption('zValue'), toOption('aValue')];
+
+const mockModel = {
+  _getKeys: jest.fn().mockReturnValue(mockKeys),
+  _getOperators: jest.fn(() => [{ label: '=', value: '=' }, { label: '!=', value: '!=' }]),
+  _getValuesFor: jest.fn().mockReturnValue(mockValues),
+  _updateFilter: jest.fn(),
+  _removeFilter: jest.fn(),
+  state: { filters: [], readOnly: false },
+};
+
+describe('FilterRenderer Component', () => {
+  let user: ReturnType<typeof userEvent.setup>;
+
+  beforeEach(() => {
+    // Need to use delay: null here to work with fakeTimers
+    // see https://github.com/testing-library/user-event/issues/833
+    user = userEvent.setup({ delay: null });
+    
+    jest.clearAllMocks();
+  });
+
+  it('renders keys correctly', async () => {
+    await act(async () => {
+      render(<FilterRenderer filter={mockFilter} model={mockModel as unknown as FilterByVariable} />);
+    });
+    
+    await user.click(screen.getByText('testKey'));
+    expect(screen.getByText('kind')).toBeInTheDocument();
+    expect(screen.getByText('resource.tag1')).toBeInTheDocument();
+    expect(screen.getByText('resource.tag2')).toBeInTheDocument();
+    expect(screen.getByText('span.tag1')).toBeInTheDocument();
+  });
+
+  it('orders keys correctly', async () => {
+    const keys = formatKeys(mockKeys, [mockFilter], 'rate');
+    expect(keys).toEqual([
+      { label: 'kind', value: 'kind' },
+      { label: 'resource.tag1', value: 'resource.tag1' },
+      { label: 'resource.tag2', value: 'resource.tag2' },
+      { label: 'span.tag1', value: 'span.tag1' }
+    ]);
+  });
+
+  it('updates filter key on selection change', async () => {
+    await act(async () => {
+      render(<FilterRenderer filter={mockFilter} model={mockModel as unknown as FilterByVariable} />);
+    });
+    
+    await user.click(screen.getByText('testKey'));
+    expect(screen.getByText('kind')).toBeInTheDocument();
+    expect(screen.getByText('resource.tag1')).toBeInTheDocument();
+    await user.click(screen.getByText('kind'));
+    
+    expect(mockModel._updateFilter).toHaveBeenCalledWith(mockFilter, { key: 'kind' });
```

Comment:
Is it the only expectation we can have? As `model._updateFilter` is a "private" method from a 3rd-party library, could we assert on something that is in our control? Something we can observe in the document for instance?
(I'm not sure because I'm not super familiar with the code)

My concern is that if Scenes `AdHocFiltersVariable` is modified, this test will fail, and it shouldn't.

Reply:
The problem here is that I need to mock parts of the model, because creating the variable and letting the code flow as it normally would, will create issues as it flows through scenes and expects various parts of the model to be defined (and which will not update the UI correctly unless it is defined correctly), which it is not as this does not reflect the purpose of the test and requires extra setup to work correctly.

IMO, it's better to mock high level functions, where you have more control over the flow and what's returned. Then you can check that the code flows in an expected way and test accordingly.

I agree that we could test something in the document instead of testing a method outside of the app, but that actually requires not mocking the `_updateFilter` method for example, which is an issue for the reason mentioned above. I have however update this to use `jest.spyOn` for the methods I want to watch rather than mock the model itself.

In any case, I don't see a huge amount of value in not testing this method from scenes is called with the correct data as it's what we call directly in our `onChange`. Now I could add a util function that wraps `_updateFilter` in our `onChange` and test that is called with the correct input, but that's really just adding a layer of misdirection for the sake of the tests.